runtime/src/main/java/org/corfudb/runtime/view/stream/AddressMapStreamView.java

```diff
@@ -155,53 +155,74 @@ protected boolean discoverAddressSpace(final UUID streamId,
-                                        "address %s, while accessing the stream at version " +
-                                        "%s. Looking for a new checkpoint.",
-                                this, stopAddress, startAddress, trimMark,
-                                getCurrentContext().getCheckpoint().startAddress, maxGlobal);
-                        throw new TrimmedException(message);
-                    }
-                }
+                moveToReadQueue(streamAddressSpace, queue, startAddress, stopAddress, maxGlobal);
             }
         }
 
         addressCount += queue.size();
         return !queue.isEmpty();
     }
 
+    private void moveToReadQueue(final StreamAddressSpace streamAddressSpace,
+                                  final NavigableSet<Long> queue,
+                                  final long startAddress,
+                                  final long stopAddress,
+                                  final long maxGlobal) {
+
+        // Transfer discovered addresses to queue. We must limit to maxGlobal,
+        // as startAddress could be ahead of maxGlobal---in case it reflects
+        // the tail of the stream.
+        queue.addAll(streamAddressSpace.copyAddressesToSet(maxGlobal));
+
+        final long trimMark = streamAddressSpace.getTrimMark();
+
+        // No valid trim-mark has been defined,
+        // thus we cannot run into trimmed address space.
+        if (!Address.isAddress(trimMark)) {
+            return;
+        }
+
+        // Trim-mark is below or equal to the lowest address,
+        // thus we cannot run into trimmed address space.
+        if (trimMark <= stopAddress) {
+            return;
+        }
+
+        // In case we are dealing with a stream that does not have the checkpoint
+        // capability (transactional stream for example), raise the exception.
+        if (!isCheckpointCapable()) {
+            String message = String.format("getStreamAddressMap[{%s}] stream has been " +
+                    "trimmed at address %s and we are trying to access the " +
+                    "stream starting at address %s. This stream does not have " +
+                    "the checkpoint capability.", this, trimMark, stopAddress);
+            log.info(message);
+            throw new TrimmedException(message);
+        }
+
+        // Address maps might have been trimmed, hence not reflecting all updates to the stream
+        // For this reason, in the case of a valid trim mark, we must be sure this space is
+        // already resolved or loaded by a checkpoint.
+        //
+        // In case of a raw stream (a stream that is not consumed as part of an SMR object),
+        // checkpoint start address and the trim mark will never match, causing the
+        // TrimmedException to be thrown whenever trimMark > stopAddress.
+        if (isCheckpointCapable() && !isTrimCoveredByCheckpointOrLocalView(trimMark)) {
```

Comment:
`isCheckpointCapable()` becomes redundant now.

Reply:
I agree. I was planning to clean up this logic in the next patch. There is a quite a bit of cleanup that needs to be done. At this point, I want to minimize the amount of changes.